src/main/java/com/thirdcc/webapp/service/impl/TransactionServiceImpl.java

```diff
@@ -70,36 +61,54 @@ public TransactionServiceImpl(
         log.debug("Request to save Transaction : {}", transactionDTO);
-        eventService.findEventByIdAndNotCancelledStatus(transactionDTO.getEventId());
-        if (transactionDTO.getTransactionType() == TransactionType.EXPENSE && transactionDTO.getReceiptDTO() == null) {
-            throw new BadRequestException("Expense transaction required an receipt image as proof ");
-        }
-        if (transactionDTO.getReceiptDTO() != null) {
-            ReceiptDTO receiptDTO = saveReceipt(transactionDTO.getReceiptDTO());
-            transactionDTO.setReceiptId(receiptDTO.getId());
+
+        // EventID is nullable because some INCOME/EXPENSE might not be associated to event (donation & etc)
+        // - Check all required fields present
+        // - transaction_amount
+        //   - created_by
+        //   - transaction_status
+        //   - title
+        // - Create object
+        // - Save
+
+        if (transactionDTO.getEventId() != null){
+            // Check if there is existing event and the status is not cancelled
+            Event ongoingEvent = eventService.findEventByIdAndNotCancelledStatus(transactionDTO.getEventId());
+            if (ongoingEvent == null) {
+                throw new BadRequestException("Event is cancelled");
+            }
         }
-        Transaction transaction = transactionMapper.toEntity(transactionDTO);
-        transaction.setTransactionStatus(TransactionStatus.SUCCESS);
-        transaction = transactionRepository.save(transaction);
-        if (transactionDTO.getTransactionType() == TransactionType.EXPENSE) {
-//            createClaimRecord(transaction);
+
+        boolean isTransactionDTOValid = validateTransactionDTO(transactionDTO, multipartFile);
+
+        // Throw exception on invalid transactionDTO
+        if (!isTransactionDTOValid){
+            throw new BadRequestException(
+                "Transaction must include fields " +
+                    "transaction_amount, " +
+                    "transaction_status, " +
+                    "title, " +
+                    "image_link(expense) OR multipartFile(if no image_link)"
+            );
         }
-        return transactionMapper.toDto(transaction);
-    }
 
-    @Override
-    public TransactionDTO update(TransactionDTO transactionDTO) {
-        log.debug("Request to update Transaction: {}", transactionDTO);
-        Transaction transaction = transactionRepository
-            .findById(transactionDTO.getId())
-            .orElseThrow(() -> new BadRequestException("Id not found when updating transaction"));
-        if (transaction.getTransactionStatus() == TransactionStatus.CANCELLED) {
-            throw new BadRequestException("Cannot update transaction that is cancelled");
+        // Generate imageLink if required
+        if (transactionDTO.getImageLink() == null || transactionDTO.getImageLink().isEmpty()){
```

Comment:
If the transactionDTO already have imageLink and user want to replace it by uploading new file, it will not enter this condition and the new file will not be saved.
Or front end will make sure that the imageLink will always be empty if the user try to update the file?

Reply:
How to access the endpoint is frontend responsibility, 
- if they upload new file but didnt make the imageLink empty, we have no way to know the new file is the file on cloud storage

conclusion: frontend need to make sure image link is empty if with to update image

Reply:
ok